src/SarifDeferredSample/Program.cs

```diff
@@ -0,0 +1,97 @@
+﻿using Microsoft.CodeAnalysis.Sarif;
+using Microsoft.CodeAnalysis.Sarif.Readers;
+using Newtonsoft.Json;
+using System;
+using System.Diagnostics;
+using System.IO;
+
+namespace SarifDeferredSample
+{
+    // NOTE: Not being submitted. This is a sample of usage of the Deferred Collections.
+
+    internal class Program
+    {
+        private const float BytesPerMB = (float)(1024 * 1024);
+
+        private static void Main(string[] args)
+        {
+            bool deferred = true;
+
+            string filePath = args[0];
+            SarifLog log = null;
+
+            Console.WriteLine($"Loading {filePath}{(deferred ? " (deferred)" : "")}...");
+
+            Measure(() =>
+            {
+                JsonSerializer serializer = new JsonSerializer();
+                serializer.ContractResolver = (deferred ? SarifDeferredContractResolver.Instance : SarifContractResolver.Instance);
+
+                using (JsonTextReader reader = (deferred ? new JsonPositionedTextReader(filePath) : new JsonTextReader(new StreamReader(filePath))))
+                {
+                    log = serializer.Deserialize<SarifLog>(reader);
+                }
+
+                return $"Loaded {filePath} ({new FileInfo(filePath).Length / BytesPerMB} MB)";
+            });
+
+            Run run = log.Runs[0];
+            Measure(() =>
+            {
+                int messageLengthTotal = 0;
+
+                foreach (Result result in run.Results)
+                {
+                    messageLengthTotal += result?.Message?.Text?.Length ?? 0;
+                }
+
+                return $"Enumerated {run.Results.Count:n0} Results message total {messageLengthTotal:n0}b";
+            });
+
+            Measure(() =>
+            {
+                int fileCount = 0;
+                int uriLengthTotal = 0;
+
+                if (run.Files != null)
+                {
+                    fileCount = run.Files.Count;
+
+                    foreach (string fileKey in run.Files.Keys)
+                    {
+                        FileData file = run.Files[fileKey];
+                        uriLengthTotal += file?.FileLocation?.Uri?.OriginalString?.Length ?? 0;
+                    }
+                } 
+
+                return $"Enumerated {fileCount:n0} Files, URI total {uriLengthTotal:n0}b.";
+            });
+
+            GC.KeepAlive(log);
```

Comment:
>log [](start = 25, length = 3)

Interesting. Should our `JsonSchemaToDotNet` code generator have an option to make the root class (`SarifLog`, in our case) `IDisposable`, so this is not necessary? (This is all new to me, by the way -- I had to read [The Old New Thing -- When do I need to use GC.KeepAlive](https://blogs.msdn.microsoft.com/oldnewthing/20100813-00/?p=13153) to understand this.

The downside would be that you'd _have_ to remember to dispose the log -- that is, you'd have to remember to create it in a `using` statement -- to prevent it from leaking.

Reply:
GC.KeepAlive is only used here to make the memory use measurements accurate. Otherwise, release builds will sometimes optimize away things it knows the downstream code isn't using.